### PR TITLE
[clang-tidy] Add bugprone-suspicious-stringview-data-usage check

### LINKED ISSUE #80854: [Clang-Tidy] Create bugprone-string-view-data-usage check
I'm adding this here so, i wouldn't forget.
I run into this issue personally.

Background:
```
void something(std::string s);

int main()
{
   std::string_view sv("something");
   sv = sv.substr(0, 5);
   something(sv.data());
}
```

Because constructor in std::string that take std::string_view is implicit, this forces developer to be more ... productive.
Problem is that in such case .data() may not be null terminated or may not point to proper string.

Enforce:
Find all calls to std::string_view::data, that are passed to method/function, where other parameter isnt an std::string_view::length/size.

### PR DESCRIPTION
**Pull request by PiotrZSL**
This check identifies suspicious usages of std::string_view::data() that could lead to reading out-of-bounds data due to inadequate or incorrect string null termination.

Closes #80854

### CODE REVIEW

**Review by HerrCai0907 — APPROVED** [review]
LGTM

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+bool SuspiciousStringviewDataUsageCheck::isLanguageVersionSupported(
+    const LangOptions &LangOpts) const {
+  return LangOpts.CPlusPlus;
+}
+
+std::optional<TraversalKind>
+SuspiciousStringviewDataUsageCheck::getCheckTraversalKind() const {
+  return TK_AsIs;
+}
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:43]: Aren't these normally in the header?

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      on(ignoringParenImpCasts(
+          matchers::isStatementIdenticalToBoundNode("self"))));
+
+  Finder->addMatcher(
+      cxxMemberCallExpr(
+          on(ignoringParenImpCasts(expr().bind("self"))), callee(DataMethod),
+          expr().bind("data-call"),
+          hasParent(expr(anyOf(
+              invocation(
+                  expr().bind("call"), unless(cxxOperatorCallExpr()),
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:71]: Are you intending to exclude  `operator()` here? If so, why? (Same goes for the `AncestorCall` matcher)

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                      ignoringParenImpCasts(equalsBoundNode("data-call"))),
+                  unless(hasAnyArgument(ignoringParenImpCasts(SizeCall))),
+                  unless(hasAnyArgument(hasDescendant(expr(
+                      SizeCall,
+                      hasAncestor(expr(AncestorCall).bind("ancestor-size")),
+                      hasAncestor(expr(equalsBoundNode("call"),
+                                       equalsBoundNode("ancestor-size"))))))),
+                  hasDeclaration(namedDecl(
+                      unless(matchers::matchesAnyListedName(AllowedCallees))))),
+              initListExpr(expr().bind("init"),
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:82]: `init` is unused

**Review by 5chmidti — COMMENTED** [review]
Overall looks good

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp (hunk 1 above)
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:43]: Yes they are, but those are virtual functions, not point to put them in header, we don't gain anything.

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp (hunk 2 above)
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:71]: To be honest lot of stuff need to be implemented with operators, like:
```
 s = sv.data();
```

Consider this an simplification in version 1.0, other thing is that support for binary operators also need to be implemented.

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp (hunk 1 above)
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:43]: I don't mind either way, this was just a consistency thing

File: clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp (hunk 2 above)
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/SuspiciousStringviewDataUsageCheck.cpp:71]: Ah, okay.

**Review by 5chmidti — APPROVED** [review]
Found something unused, otherwise LGTM

**Pull request merged**